snuba/clusters/cluster.py

```diff
@@ -36,17 +38,28 @@ def get_storage_set_keys(self) -> Set[StorageSetKey]:
     def get_reader(self) -> Reader[TQuery]:
         raise NotImplementedError
 
+    @abstractmethod
+    def get_writer(
+        self,
+        table_name: str,
+        encoder: Callable[[WriterTableRow], bytes],
+        options: Optional[Mapping[str, Any]],
```

Comment:
I am not sure this should be in the abstract version of the Cluster and not in the clickhouse specific one? 
Clickhouse query settings are passed to the reader at the time we run the query, so the get_reader method does not need to know them. Do you plan to move twards that direction for the writer as well ?

Reply:
I don't have a strong opinion on whether options should be passed later at write time. Either way I think it is not really symmetric with the reader since TableWriter is not the actual writer. In an earlier variant of this PR I deleted the table writer class entirely and just returned the BatchWriter from this method directly, but am not sure it's a good idea. Or perhaps we can introduce an equivalent TableReader class that gets passed the options like this.

As for the abstract cluster class, well it might not need options but I left the options in there to keep mypy happy.

Reply:
About the idea to remove TableWriter, the main reason I reverted it is that it's a fairly invasive change, and I don't want to lump it in here. This PR aims to move in that direction though by having the storage instantiate the writer.

I still think it's worth considering merging TableWriter and WritableTableStorage since they are pretty tightly coupled. I feel like having 2 separate classes adds an unnecessary degree of indirection and adds complexity that was kind of hard to work around (like the CdcTableWriter stuff which I need to find a better solution for).